lib/pulse_counter.toit

```diff
@@ -0,0 +1,229 @@
+// Copyright (C) 2022 Toitware ApS. All rights reserved.
+// Use of this source code is governed by an MIT-style license that can be
+// found in the lib/LICENSE file.
+
+import gpio
+
+/**
+Support for the ESP32 pulse counter (PCNT).
+
+The ESP32 has up to 8 pulse-counter units, each of which has two channels.
+Channels on the same unit share a counter, but can be configured independently otherwise.
+
+# Example
+```
+import pulse_counter
+import gpio
+
+main:
+  pin := gpio.Pin 18
+  unit := pulse_counter.Unit
+  channel := unit.add_channel pin
+  while true:
+    print unit.value
+    sleep --ms=500
+```
+*/
+
+/**
+A pulse-counter channel.
+*/
+class Channel:
+  unit_/Unit
+  channel_id_/int
+  closed_/bool := false
```

Comment:
It might be nice to just use a nullable channel_id_ so that any accidental use after close is caught.